plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1363 @@
+            self._electron_temperature = T_e
+        elif T_e is None:
+            self._electron_temperature = lambda: None
+        else:
+            raise TypeError("Invalid T_e")
+
+    def electron_temperature(self, time: u.Quantity) -> u.Quantity:
+        try:
+            if not self.in_time_interval(time):
+                warnings.warn(
+                    f"{time} is not in the simulation time interval:"
+                    f"[{self.time_start}, {self.time_max}]. "
+                    f"May be extrapolating temperature."
+                )
+            T_e = self._electron_temperature(time.to(u.s))
+            if np.isnan(T_e) or np.isinf(T_e) or T_e < 0 * u.K:
+                raise NEIError(f"T_e = {T_e} at time = {time}.")
+            return T_e
+        except Exception as exc:
+            raise NEIError(
+                f"Unable to calculate a valid electron temperature " f"for time {time}"
+            ) from exc
+
+    @property
+    def n_input(self) -> u.Quantity:
+        """The number density factor input."""
+        if "H" in self.elements:
+            return self._n_input
+        else:
+            raise ValueError
+
+    @n_input.setter
+    def n_input(self, n: u.Quantity):
+        if isinstance(n, u.Quantity):
+            try:
+                n = n.to(u.cm ** -3)
+            except u.UnitConversionError:
+                raise u.UnitsError("Invalid hydrogen density.")
+            if n.isscalar:
+                self._n_input = n
+                self.hydrogen_number_density = lambda time: n
+            else:
+                if self._time_input is None:
+                    raise TypeError("Must define time_input prior to n for an array.")
+                time_input = self.time_input
+                if len(time_input) != len(n):
+                    raise ValueError("len(n) is not equal to len(time_input).")
+                f = interpolate.interp1d(
+                    time_input.value,
+                    n.value,
+                    bounds_error=False,
+                    fill_value="extrapolate",
+                )
+                self._hydrogen_number_density = lambda time: f(time.value) * u.cm ** -3
+                self._n_input = n
+        elif callable(n):
+            if self.time_start is not None:
+                try:
+                    n(self.time_start).to(u.cm ** -3)
+                    n(self.time_max).to(u.cm ** -3)
+                except Exception:
+                    raise ValueError("Invalid number density function.")
+            self._n_input = n
+            self._hydrogen_number_density = n
+        elif n is None:
+            self._hydrogen_number_density = lambda: None
+        else:
+            raise TypeError("Invalid n.")
+
+    def hydrogen_number_density(self, time: u.Quantity) -> u.Quantity:
+        try:
+            time = time.to(u.s)
+        except (AttributeError, u.UnitsError):
+            raise NEIError("Invalid time in hydrogen_density")
```

Comment:
We could probably use another validator here. And a docstring, too.

Reply:
- [ ] Create issue on applying validators in `NEI` class